### ISSUE TITLE
Non-standard dividers for YAML front matter

### ISSUE THREAD

**Issue by Giuseppe-Mazzapica** [mnapoli/FrontYAML#5]
The only package that uses FrontYAML that I know is [Couscous](http://couscous.io/).

What I like much of that app is the easiness to manage both *raw* repo .md files and GitHub pages site.

However, if some front matter is used in .md files, it's not rendered as HTML in GH pages, but of course it is in repo files if read from GitHub.

For example look at the GitHub [readme of Couscous](https://github.com/CouscousPHP/Couscous/blob/master/README.md) itself and you'll see the front matter rendered as one column HTML table on top of file.

To be honest, I don't like that.

For that reason, I forked FrontMatter and **edited the `Parser::parse()` method to accept non-standard dividers for front matter**.

Here is the [diff comparison](https://github.com/Giuseppe-Mazzapica/FrontYAML/compare/alt-separators).

After that I forked Couscous too, and edited `ParseMarkdownFrontMatter` class to use the edited FrontYAML `Parser::parse()` method using HTML comment syntax as front matter separators. 
Something like

```php
$document = $this->markdownParser->parse($file->getContent(), true, '<!--', '-->');
```
In this way front matter for my .md files now is something like:

```html
<!--
layout: "home"
currentMenu: "welcome"
-->
```
As a result **front matter is invisible** (because rendered as HTML comment) when *raw* .md files are read from GitHub, but it is perfectly parsed by Couscous.

My Couscous fork is not public at the moment, because I'd like to make front matter separator configurable via `couscous.yml`, but no time for that.
BTW that's better discussed as Couscous issue.

## Question

Do you want I make a PR to allow custom dividers for front matter as seen in [my fork](https://github.com/Giuseppe-Mazzapica/FrontYAML/compare/alt-separators)?

**Comment by Rarst** [comment]
+1 for HTML comment approach. I evaluated bunch of things when making my small static parser for own use. Comments approach greatly reduces header showing up when you don't want it.

**Comment by Narno** [comment]
:+1:

**Comment by mnapoli** [comment]
I love that!

> Do you want I make a PR

Hell yeah ;)

Maybe it would even be possible to match both `---` and HTML comments? I think I would like to use comments for Couscous too, but I don't really want to make it yet another configuration option. It should support both separators out of the box. Maybe it's a Couscous problem but it could be solved here?

**Comment by Giuseppe-Mazzapica** [comment]
The approach currently used in my fork allows to use any kind of separator. You can even do crazy things like

```
°º¤ø,¸¸,ø¤º°`°º¤ø,¸,ø¤°º¤ø,¸¸,ø¤º°`°º¤ø
layout: "home"
currentMenu: "welcome"
°º¤ø,¸¸,ø¤º°`°º¤ø,¸,ø¤°º¤ø,¸¸,ø¤º°`°º¤ø
```

and IMHO for what regards FrontYAML is better keep this approach, leaving to package consumers the choice of separator to use (keeping default to `---`).

For what regards Couscous, to match both HTML comments and `---` is possible only if I remove `preg_quote` from my current code, in this way is possible to use custom regex for separators in Couscous.

This is a bit *dangerous* because is possible to break the whole regex by using bad arguments, but if the separators regex will be hardcoded in Couscous that shouldn't be a problem.

If I do that, then in Couscous is possible to use a regex that matches both separators, something like:

```php
$document = $this->markdownParser->parse($file->getContent(), true, '<!---|---', '-->|---');
```

In summary, if you are agree I can:

 1. edit current code in my FrontYAML fork to allow custom regex for separators, then send a PR
 2. publish my fork of Couscous after made HTML comment and `---` both supported out of the box (using method described above), then send a PR to Couscous

**Comment by mnapoli** [comment]
I have a few suggestions if that's fine by you:

- pass these options in the constructor, because those are not the kind of options you would change for each file (keeping `parse()` simple would be great)
- default values would be `---` (to keep BC with current behavior)
- escape the delimiter strings with [`preg_quote()`](http://php.net/manual/fr/function.preg-quote.php) so that we can pass delimiters without having to know that it's going to be used in a regex (encapsulation of the internal implementation)
- instead of passing `<!--|---` leave the possibility to pass a string or an array, so that we can pass `---` or `array('---', '<!--')`: then you can do `implode('|', array_map('preg_quote', $left))` or something similar to build the regex

If you have time to add a test that's awesome, else I'll do it don't worry.

Thanks!

**Comment by Giuseppe-Mazzapica** [comment]
> pass these options in the constructor, because those are not the kind of options you would change for each file (keeping parse() simple would be great)

Not really a fun of this. Actually there is a single constructor call just like a single parse call, so options are set once in both cases. But I usually prefer to keep methods as stateless as possible. But if you prefer that way absolutely no problem for me.

> default values would be --- (to keep BC with current behavior)

Of course. Mentioned that in my previous comment too.

>escape the delimiter strings with preg_quote() so that we can pass delimiters without having to know that it's going to be used in a regex (encapsulation of the internal implementation)

That's the reason my current solution uses `preg_quote`

> instead of passing <!--|--- leave the possibility to pass a string or an array, so that we can pass --- or array('---', '<!--'): then you can do implode('|', array_map('preg_quote', $left)) or something similar to build the regex

Clever solution! let's go with that.

I'll implement these changes today evening or tomorrow.

**Comment by mnapoli** [comment]
Awesome thanks!

> Not really a fun of this. Actually there is a single constructor call just like a single parse call, so options are set once in both cases. But I usually prefer to keep methods as stateless as possible. But if you prefer that way absolutely no problem for me.

I would consider those configuration rather than state. E.g. in Couscous this would be a global configuration of the front yaml parser, not something that would change from file to file.

In a repository containing 10+ markdown files, the front yaml parser would use the same separators for every file (constructor called 1 time, `parse()` called 10 times).

What do you think?

**Comment by Giuseppe-Mazzapica** [comment]
> I would consider those configuration rather than state.

Technically speacking it's state :) A method is stateless if given same arguments return same result. If we set configuration in constructor `parse()` may return different things even if same arguments are provided, it depends on instance *state*, so...

> In a repository containing 10+ markdown files, the front yaml parser would use the same separators for every file (constructor called 1 time, parse() called 10 times).

I think that a lot depends on what consumer whan to do.

If we think as Couscous as the sole consumer, than the 2 methods are equivalent: even if `parse()` can be actually called dozen of times, there will be just one occurrence of it in the code.

If a consumer wants/needs to allow different separators for different files, pass data to method is better.

On the countrary, if the consumer want to force just one separator the constructor method make more sense. 

So IMHO it's really all about preference and, as said, I prefer stateless methods when possible.

But once I understand you prefer constructor approach in this case it's perfectly fine for me.

PR done :)

**Comment by mnapoli** [comment]
Yeah indeed it depends on the use case. Thanks for doing that, I'm releasing!
